src/java.net.http/share/classes/jdk/internal/net/http/HttpResponseImpl.java

```diff
@@ -95,6 +98,15 @@ public T body() {
         return body;
     }
 
+    @Override
+    public Optional<T> bodyWhen(Predicate<ResponseInfo> predicate) {
+        if (predicate.test(rinfo)) {
+            return Optional.of(body());
+        } else {
+            return Optional.empty();
```

Comment:
should close the body here if it is closeable - or subscribe and cancel the subscribtion if it is a publisher.

Reply:
See https://git.openjdk.org/jdk/pull/18270
and https://bugs.openjdk.org/browse/JDK-8308364?focusedId=14583793&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-14583793